src/doc/rustc/src/platform-support/armv6k-nintendo-3ds.md

```diff
@@ -0,0 +1,104 @@
+# `armv6k-nintendo-3ds`
+
+**Tier: 3**
+
+The Nintendo 3DS platform, which has an ARMv6K processor, and its associated
+operating system (`horizon`).
+
+## Target maintainers
+
+- [@Meziu](https://github.com/Meziu)
+- [@AzureMarker](https://github.com/AzureMarker)
+- [@ian-h-chamberlain](https://github.com/ian-h-chamberlain)
+
+## Requirements
+
+This target is cross-compiled. `#![no_std]` crates should work without much
+additional effort.
+
+`std` is partially supported, but mostly works. Some APIs are unimplemented
+and will simply return an error, such as `std::process`. An allocator is provided
+by default.
+
+In order to support `std` APIs, binaries must be linked against the open-source
+[devkitARM toolchain](https://devkitpro.org/wiki/Getting_Started), notably
+using its `arm-none-eabi-gcc` as the linker.
+
+Additionally, some helper crates provide implementations of some `libc` functions
+use by `std`. These, or an alternate implementation of the relevant functions,
+are required to use `std`:
+
+- [`pthread-3ds`](https://github.com/Meziu/pthread-3ds) provides pthread APIs for `std::thread`.
+- [`linker-fix-3ds`](https://github.com/Meziu/rust-linker-fix-3ds) fulfills some other missing libc APIs.
+
+Binaries built for this target should be compatible with most variants of the
+3DS (and 2DS) hardware and firmware, but testing is limited and some versions may
+not work correctly.
+
+This target generates binaries in the ELF format.
+
+## Building the target
+
+You can build Rust with support for the target by adding it to the `target`
```

Comment:
Actually I should test this too... I have built `std` for the target like this but never did an end-to-end since `cargo 3ds` is pretty convenient and the `std` code was changing pretty often.

Reply:
I ran into some small issues with this. At first just some warnings-turned-errors, which I fixed in https://github.com/AzureMarker/rust-horizon/commit/96295e829409c6ac4a8d92b2bbfd5bbfe0464694 and https://github.com/AzureMarker/rust-horizon/commit/b51f37d1611a156d1333593112bf15fb4b080d5c. But then actually compiling something would give me the following link error (simplified paths a bit):
```
/opt/devkitpro/devkitARM/arm-none-eabi/bin/ld: error: hello_world-9db584d0ec6e1906.elf uses VFP register arguments, libcompiler_builtins-e5be153354215cbd.rlib(aeabi_div0.o) does not
          /opt/devkitpro/devkitARM/arm-none-eabi/bin/ld: failed to merge target specific data of file libcompiler_builtins-e5be153354215cbd.rlib(aeabi_div0.o)
```

Reply:
I forgot about the warnings-as-errors (we were going to have to fix those to merge upstream anyway, so thanks for fixing!).

As for the linker errors, I saw the same while trying to get `./x.py test` to build for the target, which makes me think if we could figure out how to solve it here we would be closer on that front as well. I asked on Zulip if anyone knew how to help, but unfortunately got no responses... 

I had also seen this kind of link error when hacking around previously (maybe when working on https://github.com/Meziu/libc/pull/8 ?), but I cannot quite remember how I got past it back then. I did find this issue, but it's a bit old and doesn't seem 100% relevant for our scenario, since `build-std` works correctly: https://github.com/rust-embedded/book/issues/255

Also found [this stackoverflow](https://stackoverflow.com/questions/9753749/arm-compilation-error-vfp-registers-used-by-executable-not-object-file), which is jogging my memory that perhaps I exported `CFLAGS` or `CXXFLAGS` (or both) somewhere to get around it (forcing to build with the same flags as in https://github.com/AzureMarker/rust-horizon/blob/horizon-std/compiler/rustc_target/src/spec/armv6k_nintendo_3ds.rs . 

`libc-test` uses the `cc-rs` crate, though, which may not be applicable for `compiler_builtins`, so maybe something else is at play, but that might be worth a try? Or perhaps there's some `config.toml` values that can be tweaked to get the behavior needed here?

Reply:
Yeah the `VFP register arguments` is something I had to fix too. Putting those vars in the target specification fixed it. I believe it has to do with the size/instructions available to the CPU, which the other linking processes (like `libctru` or whatever happens in `arm-none-eabi-gcc`) don’t find compatible in some specific cases. I even tried to get down to only one of those flags being the problem, though I don’t remember which, and I judged it’d be better to keep all of them anyways, for future problems may arise.

Reply:
> ```shell
> $ export CFLAGS_armv6k_nintendo_3ds='-mfloat-abi=hard -mtune=mpcore -mtp=soft -march=armv6k'
> ```

Setting this before `./x.py build` seems to work. I didn't even set it before `cargo build --target armv6k-nintendo-3ds --example hello-world` in ctru-rs and it worked (separate terminal). I was able to execute it just fine on my 3DS, after converting to 3dsx.

Reply:
Nice! I will add to the docs about `Building Rust for the Target`. I think after digging a bit more, this is a gap in the default flags set by `cc-rs`, since it doesn't have knowledge about the target architecture. We could probably deliver a fix upstream there to resolve it, but exporting `CFLAGS_armv6k_nintendo_3ds` seems like the best workaround meanwhile.